k-es-grpc/src/main/kotlin/no/ks/kes/grpc/GrpcSubscriptionListener.kt

```diff
@@ -0,0 +1,90 @@
+                                private val metadataSerdes: EventMetadataSerdes<out Metadata>? = null
+) : SubscriptionListener() {
+
+    override fun onEvent(subscription: Subscription, resolvedEvent: ResolvedEvent) {
+
+        log.debug { "$hwmId: received event \"$resolvedEvent\"" }
+
+        val eventNumber = resolvedEvent.originalEvent.revision
+
+        subscriptionLiveCheckpoint.triggerOnceIfSubscriptionIsLive(eventNumber) {
+            onLive.invoke()
+        }
+
+        when {
+            !resolvedEvent.isResolved() ->
+                log.info { "$hwmId: event not resolved: ${resolvedEvent.link.revision} ${resolvedEvent.link.streamId}" }
+
+            resolvedEvent.isIgnorable() ->
+                log.info { "$hwmId: event ignored: ${resolvedEvent.originalEvent.revision} ${resolvedEvent.originalEvent.streamId}" }
+
+            else -> try {
+                val eventMeta =
+                    if (resolvedEvent.event.userMetadata.isNotEmpty() && metadataSerdes != null) metadataSerdes.deserialize(
+                        resolvedEvent.event.userMetadata
+                    ) else null
+                val event = EventUpgrader.upgrade(
+                    serdes.deserialize(
+                        resolvedEvent.event.eventData,
+                        resolvedEvent.event.eventType
+                    )
+                )
+                val aggregateId = UUID.fromString(resolvedEvent.event.streamId.takeLast(36))
+                onEvent.invoke(
+                    EventWrapper(
+                        Event(
+                            aggregateId = aggregateId,
+                            metadata = eventMeta,
+                            eventData = event
+                        ),
+                        eventNumber = eventNumber,
+                        serializationId = serdes.getSerializationId(event::class)
+                    )
+                )
+                    .also {
+                        log.info(
+                            "$hwmId: event ${eventNumber}@${resolvedEvent.originalEvent.streamId}: " +
+                                    "${resolvedEvent.event.eventType}(${resolvedEvent.event.eventId}) received"
+                        )
+                    }
+            } catch (e: java.lang.Exception) {
+                log.error(e) { "Event handler for $hwmId threw exception: " }
+                throw e
+            }
+        }
+
+        lastEventProcessed.set(eventNumber)
+    }
+
+    override fun onCancelled(subscription: Subscription?) {
+        log.info { "Subscription cancelled: subscriptionId=${subscription?.subscriptionId}, subscriber=$hwmId, streamId=$streamId, lastEvent=$lastEventProcessed" }
+    }
+
+    override fun onError(subscription: Subscription?, throwable: Throwable?) {
+        log.error { "error on subscription. subscriptionId=${subscription?.subscriptionId}, subscriber=$hwmId, streamId=$streamId, lastEvent=$lastEventProcessed, exception=$throwable" }
```

Comment:
subscriber=$hwmId -> hwmId=$hwmId?

log.error(throwable) {...} i stedet for å logge den i meldingen?